### DISCUSSION TITLE
New Keycloak Realm not providing roles for app login

### DISCUSSION THREAD

**Discussion by GregJohnStewart** [keycloak/keycloak#22398] [Q&A] [answered]
Also posted here: https://stackoverflow.com/questions/76879100/new-keycloak-realm-not-providing-roles-for-app-login

I'm working on setting up Keycloak to be my auth provider, using OIDC. My app is written using Quarkus.

I am creating a new realm using the following steps:

1. Enter Keycloak Admin Console
2. Under realm dropdown, hit "Create Realm". Name it `oqm`.
3. Realm Settings -> General
    1. Realm ID changed to `oqm`
    2. Display name and HTML Display name set to `Open QuarterMaster`
    3. User-managed access turned on
    4. Hit "Save"
4. Realm Settings -> Login. Turn on:
    - User Registration
    - Forgot Password
    - Remember Me
    - Email as Username?
    - Edit Username
5. Authentication -> Policies -> Password Policy. Add the following policies:
    - Minimum Length of 8
    - Not username
    - Not email
    - Special chars
    - Uppercase letters
    - Lowercase letters
    - Digits
5. Create Client
    1. Clients -> Create Client
    2. Client id: `oqm-app`
    3. Name: `Open QuarterMaster App`
    4. Always display in UI
    5. Client authentication on
    6. "Standard flow", "Direct access grants", "Implicit flow", and "Service Account Roles" all on
    7. Valid redirect URIs: `*`
    8. Valid post logout redirect URIs: `+`
    9. Create
6. Add roles
   - `inventoryAdmin` -> `Role to enable inventory administration. Can import/export inventory data.`
   - `inventoryView` -> `Role to enable viewing inventory.`
   - `inventoryEdit` -> `Role to enable editing inventory.`
   - `itemCheckout` -> `Role to enable checking out (and back in) items.`
7. Add groups with roles:
   - `users`
       - `inventoryView`
       - `inventoryEdit`
       - `itemCheckout`
   - `admins` as child group of `users`
       - `inventoryAdmin`
8. Realm Settings -> User Registration -> Default Groups
   1. Add `admins` to default group (I have also tried adding all roles as default)

However, when I create a user and login, the app returns a 403, with the following log output:

```
12:53:07 [DEBUG][9202a4ca6299115508ca0c543b9beb76|  |c06abf9f49a51772|true ][io.qu.oi.ru.DefaultTenantConfigResolver |vert.x-eventloop-thread-3 ]():: Registered TenantResolver has not provided the configuration for tenant 'rt', using the default tenant
12:53:07 [DEBUG][9202a4ca6299115508ca0c543b9beb76|  |c06abf9f49a51772|true ][io.qu.oi.ru.CodeAuthenticationMechanism |vert.x-eventloop-thread-3 ]():: Session cookie is present, starting the reauthentication
12:53:07 [DEBUG][9202a4ca6299115508ca0c543b9beb76|  |c06abf9f49a51772|true ][io.qu.oi.ru.DefaultTenantConfigResolver |vert.x-eventloop-thread-3 ]():: Registered TenantResolver has not provided the configuration for tenant 'rt', using the default tenant
12:53:07 [DEBUG][9202a4ca6299115508ca0c543b9beb76|  |c06abf9f49a51772|true ][io.qu.oi.ru.OidcIdentityProvider        |vert.x-eventloop-thread-3 ]():: Starting creating SecurityIdentity
12:53:07 [DEBUG][9202a4ca6299115508ca0c543b9beb76|  |c06abf9f49a51772|true ][io.qu.oi.ru.DefaultTenantConfigResolver |vert.x-eventloop-thread-3 ]():: Registered TenantResolver has not provided the configuration for tenant 'rt', using the default tenant
12:53:07 [DEBUG][9202a4ca6299115508ca0c543b9beb76|  |c06abf9f49a51772|true ][io.qu.oi.ru.OidcIdentityProvider        |vert.x-eventloop-thread-3 ]():: Verifying the JWT token with the local JWK keys
12:53:07 [DEBUG][9202a4ca6299115508ca0c543b9beb76|  |c06abf9f49a51772|true ][io.qu.oi.ru.OidcUtils                   |vert.x-eventloop-thread-3 ]():: No claim exists at the path 'groups' at the path segment 'groups'
12:53:07 [DEBUG][9202a4ca6299115508ca0c543b9beb76|  |c06abf9f49a51772|true ][io.qu.oi.ru.OidcUtils                   |vert.x-eventloop-thread-3 ]():: No claim exists at the path 'realm_access/roles' at the path segment 'realm_access'
12:53:07 [DEBUG][9202a4ca6299115508ca0c543b9beb76|  |c06abf9f49a51772|true ][io.qu.oi.ru.OidcUtils                   |vert.x-eventloop-thread-3 ]():: No claim exists at the path 'resource_access/oqm-app/roles' at the path segment 'resource_access'
```

It seems to me that the roles never make it into the jwt? I am modeling my realm after the one provided by Quarkus' keycloak dev service, which works when I add the roles required to the users provided there.

Content of "KEYCLOAK_IDENTITY" from jwt.io:

```
{
  "exp": 1691800027,
  "iat": 1691764027,
  "jti": "ef4a49f0-db0d-49f8-a2a2-36bd35cdb4ef",
  "iss": "http://localhost:32769/realms/oqm",
  "sub": "a1f2f54d-35b9-4e94-b673-0fbec79d7c33",
  "typ": "Serialized-ID",
  "session_state": "08730e5a-7f3b-4731-8062-c0083bfef614",
  "sid": "08730e5a-7f3b-4731-8062-c0083bfef614",
  "state_checker": "5F2Y4AL-QC1KUOwTLBXEGK_ZrHdlq9na6d0_joCJTaQ"
}
```

I would think I should see username/email/name as well as roles, but none are present? This is the only cookie that resembles a jwt, except maybe `q_session*`, but it is split up. I am wondering about the "two token" thing, I was expecting to see two jwt's, one identity and one auth?

I can obviously sign in fine, as going to keykloak's account page shows my user's info, just not have the right info passed to my app in the tokens.

**Comment by sschu** [comment, marked as answer]
You can use the admin console the check how the token looks like (go to the client, tab Client Scope, Evaluate). You might have tweak the client configuration so it is actually allowed to get the roles into the token that the user has (see https://www.keycloak.org/docs/22.0.1/server_admin/#_role_scope_mappings).

**Reply by GregJohnStewart** [reply]
Thanks for the reply, everything looks as it should though? If anything, my new realm's client seems to have more scopes associated with it:
![image](https://github.com/keycloak/keycloak/assets/7083701/10a26fd3-4c3e-44cb-a28b-656aadc8d75e)

I also made sure that the "Full scope allowed" is on, as per the link you sent, which it already was

**Reply by GregJohnStewart** [reply]
![image](https://github.com/keycloak/keycloak/assets/7083701/8ff9bca6-3a60-4ffb-92f4-3af6e950670b)

**Reply by GregJohnStewart** [reply]
Upon looking at the evaluated token from the user selected, it looks like the roles aren't being placed in a "groups" claim:

![image](https://github.com/keycloak/keycloak/assets/7083701/54f301f9-c340-4874-b8b6-aef944f2c4d4)

JWT from the quarkus realm:

![image](https://github.com/keycloak/keycloak/assets/7083701/ef6336c8-550d-4429-a0cc-8cc47eb54f2e)

**Reply by GregJohnStewart** [reply]
It is also odd to me that this does not seem to be the same token that is present in the `KEYCLOAK_IDENTITY` cookie, and that that jwt doesn't match what is shown here?

Just confirmed that I only get one JWT, even from the Quarkus equivalent, in that cookie:

```
{
  "exp": 1691776510,
  "iat": 1691775910,
  "jti": "e4d968c6-08fc-46a1-ad54-5696ec2bd208",
  "iss": "http://localhost:32789/realms/quarkus",
  "sub": "02c0cd43-dfaa-4bed-8eea-584c53ae1ef6",
  "typ": "Serialized-ID",
  "session_state": "a066b1ef-e421-448f-bb91-0b0085928424",
  "sid": "a066b1ef-e421-448f-bb91-0b0085928424",
  "state_checker": "oAco5XFnYSQlMseH69KZ82kITLz3M4MgW-4fvnlVmxM"
}
```

**Reply by GregJohnStewart** [reply]
So fairly confused at this point as to how this works and where things come from and what is supposed to be where. Maybe the access token is automatically retrieved at each page load?

**Reply by GregJohnStewart** [reply]
Looks like that the token that is eventually passed to the app as the `id` token, and in fact exists, and isn't being held as a cookie, but does exist and looks like it should...:

```
{
  "exp": 1691779564,
  "iat": 1691778964,
  "auth_time": 1691778964,
  "jti": "28f6102a-5287-424f-987d-0d91a1f26357",
  "iss": "http://localhost:32789/realms/quarkus",
  "aud": "quarkus-app",
  "sub": "02c0cd43-dfaa-4bed-8eea-584c53ae1ef6",
  "typ": "ID",
  "azp": "quarkus-app",
  "session_state": "46072f1a-2623-4f57-a241-3285c9f882ed",
  "at_hash": "k0eCj4mIdDkj_yF6IiQt5w",
  "sid": "46072f1a-2623-4f57-a241-3285c9f882ed",
  "upn": "alice",
  "groups": [
    "inventoryView",
    "admin",
    "user"
  ]
}
```

**Reply by GregJohnStewart** [reply]
Figured it out, needed to: Clients -> oqm-app -> Client Scopes, set `microprofile-jwt` to "default"

Thanks for the eventual point to the right place
